SDMA: Fix sound issues caused by SDMA driver

dbaluta opened the pull request:
This patch series:
* introduces sdma channel states in order to prevent invalid state transitions
* adds a fix for pause/resume sound issue because computation for total buffer length was wrong.
* adds a fix for a sound issue caused by a race when updating buffer status

dbaluta committed:
drivers: dma: sdma: Fix noise issue with pause/resume

Each time we configure an SDMA channel we also compute the total
allocated DMA buffer length but we assume is initialized with zero.

This is true each time a channel is requested. But if there are
multiple calls to configure without releasing the channel the buffer
length is not correctly computed.

So, we need to initialize it with zero each time we reconfigure the dma
channel.

Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

dbaluta commented:
@iuliana-prodan  please review.

LaurentiuM1234 commented on drivers/dma/dma_nxp_sdma.c:
```diff
@@ -395,6 +423,12 @@ static int dma_nxp_sdma_reload(const struct device *dev, uint32_t channel, uint3
 
+	/* allow reload only for active channels */
+	if (chan_data->state != SDMA_CHAN_STATE_STARTED) {
+		LOG_ERR("%s: invalid state %d", __func__, chan_data->state);
```
drop the function name prefix. `CONFIG_LOG_FUNC_NAME_PREFIX_` should already do the job.

LaurentiuM1234 commented on drivers/dma/dma_nxp_sdma.c:
```diff
@@ -347,8 +365,15 @@ static int dma_nxp_sdma_start(const struct device *dev, uint32_t channel)
 
+	if (chan_data->state != SDMA_CHAN_STATE_STOPPED &&
+	    chan_data->state != SDMA_CHAN_STATE_CONFIGURED) {
+		LOG_ERR("%s: invalid state %d", __func__, chan_data->state);
```
same here

LaurentiuM1234 commented on drivers/dma/dma_nxp_sdma.c:
```diff
@@ -20,6 +20,15 @@ LOG_MODULE_REGISTER(nxp_sdma);
+	SDMA_CHAN_STATE_STARTED,
+	SDMA_CHAN_STATE_STOPPED,
+	SDMA_CHAN_STATE_SUSPENDED,
+	SDMA_CHAN_STATE_RELEASING,
```
seems unused?

LaurentiuM1234 commented on drivers/dma/dma_nxp_sdma.c:
```diff
@@ -20,6 +20,15 @@ LOG_MODULE_REGISTER(nxp_sdma);
+	SDMA_CHAN_STATE_CONFIGURED,
+	SDMA_CHAN_STATE_STARTED,
+	SDMA_CHAN_STATE_STOPPED,
+	SDMA_CHAN_STATE_SUSPENDED,
```
seems unused?

dbaluta replied:
@LaurentiuM1234, added them to match the states in the edma driver. I think prefer to keep them for that reason. Likely use them when add PM support.

dbaluta committed:
drivers: dma: sdma: Make access to DMA channel stats atomic

DMA channel stats like pending_length or free is not protected
and can be modified in parallel by a consumer and a producer.

This can result in non-atomic updates which in turn will result
in using stale data.

Fix this by making regions of code accessing dma stats atomic.

Fixes: e94c86f3951 ("drivers: dma: Add initial support for NXP SDMA")
Signed-off-by: Daniel Baluta <daniel.baluta@nxp.com>

teburd reviewed (changes requested):
Er meant to request changes, the state machine doesn't quite match the documented expectations

teburd commented:
> @teburd I removed the patch that handles the states for now. Lets merge the 2 bugfixes. Will need to study more the DMA state changes diagram.
> 
> In the DMA state diagram looks like you cannot call directly dma_stop() while in RUNNING state which is odd.
> 
> I will come back the states when implementing PM support for SDMA driver.
> 
> ![image](https://private-user-images.githubusercontent.com/62316/448455439-ef67c838-9fee-4bd5-a3fc-364600378f88.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg0NTI1MTgsIm5iZiI6MTc0ODQ1MjIxOCwicGF0aCI6Ii82MjMxNi80NDg0NTU0MzktZWY2N2M4MzgtOWZlZS00YmQ1LWEzZmMtMzY0NjAwMzc4Zjg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTI4VDE3MTAxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEwZmE3ZDc0MzY3ZjIwZWU1ZGU2ZDY0ODdiN2U0MDkxMzAyZjFjY2ZjNTQ3ZWEyYjllY2ZmMmQ1M2EwNzM3YTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0kWDHOfRvTg95bjbVWF7JZ1i4Y2EVQPDdVzDjtNDxPo)

It's there, but the diagram is a bit dense for sure
